src/data.c

```diff
@@ -0,0 +1,68 @@
+#include <stdlib.h>
+#include <string.h>
+#include "data.h"
+
+/* Função que cria um novo elemento de dados data_t e que inicializa 
+ * os dados de acordo com os argumentos recebidos, sem necessidade de
+ * reservar memória para os dados.	
+ * Retorna a nova estrutura ou NULL em caso de erro.
+ */
+struct data_t *data_create(int size, void *data){
+    
+    if (size <= 0 || data == NULL)
+        return NULL;
+
+    struct data_t* d = malloc (sizeof(struct data_t));
+
+    d->data = data;
+    d->datasize = size;
+
+    return d;
+}
+
+/* Função que elimina um bloco de dados, apontado pelo parâmetro data,
+ * libertando toda a memória por ele ocupada.
+ * Retorna 0 (OK) ou -1 em caso de erro.
+ */
+int data_destroy(struct data_t *data){
+
+    if(data == NULL)
+        return -1;
+
+    free(data);    
+
+    return 0;
+}
+
+/* Função que duplica uma estrutura data_t, reservando a memória
+ * necessária para a nova estrutura.
+ * Retorna a nova estrutura ou NULL em caso de erro.
+ */
+struct data_t *data_dup(struct data_t *data){
+
+    if(data == NULL || data->data == NULL || data->datasize <= 0)
+        return NULL;
+
+    struct data_t* dup = malloc (sizeof(struct data_t));
+    dup->datasize = data->datasize;
+    memcpy(dup->data, data->data, dup->datasize);
+
+    return dup;
+}
+
+/* Função que substitui o conteúdo de um elemento de dados data_t.
+ * Deve assegurar que liberta o espaço ocupado pelo conteúdo antigo.
+ * Retorna 0 (OK) ou -1 em caso de erro.
+ */
+int data_replace(struct data_t *data, int new_size, void *new_data){
+
+    if(data == NULL || data->data == NULL || data->datasize <= 0 || new_size <= 0 || new_data == NULL)
+        return -1;
+
+    free(data->data);
+    free(data->datasize);
```

Comment:
As far as I understand, this shouldn't work. You might've not found an issue bc your intellisense is not working yet, or bc **everything** in C is technically a number, including pointers. But data->datasize is not a pointer as required by `free()`. It's actually an `int`.

Reply:
![image](https://github.com/padrezulmiro/sd-projeto/assets/1679672/76327ed2-633d-4ed1-96b2-bf0907907a88)

Reply:
ah ofc, that was just dumb of me, yes it does not give an error, but I was aware of it, I simply forgor